### PR TITLE
Avoid iteration for singleton selections

### LINKED ISSUE #6131: Add fast-path for `select_candidate` with singletons
We already know the version, and the version maps are indexed by version. Can we just skip the iterator?

In `echo "apache-airflow[google]" | cargo run pip compile -`, about 4-5% of the calls to `select_candidate` use a singleton.

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

If we have a singleton `Range`, we don't need to iterate over the map of available ranges; instead, we can just get the singleton directly.

Closes #6131.

### CODE REVIEW

**Comment by charliermarsh** [comment]
Most of the test failures are just aesthetic, but there is one real failure due to this:

```rust
// If we have at least one stable release, we shouldn't allow the "if-necessary"
// pre-release strategy, regardless of whether that stable release satisfies the
// current range.
prerelease = Some(PrereleaseCandidate::NotNecessary);
```

So the selector actually _does_ rely on seeing versions outside the range, hmm.

File: crates/uv-resolver/src/candidate_selector.rs [hunk 1]
```diff
@@ -324,7 +324,31 @@ impl CandidateSelector {
...
+                {
+                    if version_maps.iter().any(|version_map| {
+                        version_map
+                            .versions()
+                            .all(|version| version.any_prerelease())
+                    }) {
+                        AllowPrerelease::Yes
+                    } else {
+                        AllowPrerelease::No
+                    }
```
Comment (charliermarsh) [inline crates/uv-resolver/src/candidate_selector.rs:346]: A little bit of a penalty here but should be somewhat rare.

Comment (charliermarsh) [reply crates/uv-resolver/src/candidate_selector.rs:342]: We only need to do this slow path when we see a pre-release singleton; and the check will short-circuit as soon as we see any non-pre-release.

Comment (charliermarsh) [reply crates/uv-resolver/src/candidate_selector.rs:342]: It'd be nice to avoid this though.

Comment (konstin) [reply crates/uv-resolver/src/candidate_selector.rs:342]: You could store whether all versions are prereleases in the version map, probably helpful for debugging anyway

File: crates/uv-resolver/src/version_map.rs [hunk 2]
```diff
@@ -138,24 +146,50 @@ impl VersionMap {
...
-            }
-            VersionMapInner::Lazy(ref lazy) => {
-                either::Either::Right(lazy.map.iter().map(|(version, dist)| {
-                    let version_map_dist = VersionMapDistHandle {
-                        inner: VersionMapDistHandleInner::Lazy { lazy, dist },
-                    };
-                    (version, version_map_dist)
-                }))
-            }
+        if let Some(version) = range.as_singleton() {
```
Comment (konstin) [inline crates/uv-resolver/src/version_map.rs:151]:
```suggestion
        // Performance optimization: If we only have a single version, return that version directly.
        if let Some(version) = range.as_singleton() {
```

Without a comment, i can already see myself looking at this code later and wondering whether the branch is load-bearing or not.

**Review by konstin — APPROVED** [review]

File: crates/uv-resolver/src/candidate_selector.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv-resolver/src/candidate_selector.rs:342]: Good idea.

**Pull request merged**